dm/master/server.go

```diff
@@ -1334,37 +1341,232 @@ func (s *Server) generateSubTask(ctx context.Context, task string) (*config.Task
-// grpc client, worker id (host:port), subtask config in toml, task name and error
-func (s *Server) taskConfigArgsExtractor(cfg *config.SubTaskConfig) (*scheduler.Worker, string, string, error) {
-	handleErr := func(err error) error {
-		log.L().Error("response", zap.Error(err))
-		return err
+func extractWorkerError(result *pb.ProcessResult) error {
+	if result != nil && len(result.Errors) > 0 {
+		return terror.ErrMasterOperRespNotSuccess.Generate(result.Errors[0].String())
+	}
+	return nil
+}
+
+/*
+QueryStatus for worker response
+Source:
+	OperateSource:
+		* StartSource, UpdateSource: sourceID = Source
+		* StopSource: return resp.Result = false && resp.Msg = “worker has not started”.
+		  But if len(resp.SourceStatus.Result) > 0, it means dm-worker has some error in
+		  StopWorker and it should be pushed to users.
+Task:
+	StartTask, UpdateTask: query status and related subTask stage is running
+	OperateTask:
+		* pause: related task status is paused
+		* resume: related task status is running
+		* stop: related task can't be found in worker's result
+Relay:
+	OperateRelay:
+		* pause: related relay status is paused
+		* resume: related relay status is running
+In the above situations, once we find an error in response we should return the error
+*/
+func (s *Server) waitOperationOk(ctx context.Context, cli *scheduler.Worker, taskName, sourceID string, masterReq interface{}) (*pb.QueryStatusResponse, error) {
+	var expect pb.Stage
+	switch masterReq.(type) {
+	case *pb.OperateSourceRequest:
+		req := masterReq.(*pb.OperateSourceRequest)
+		switch req.Op {
+		case pb.SourceOp_StartSource, pb.SourceOp_UpdateSource:
+			expect = pb.Stage_Running
+		case pb.SourceOp_StopSource:
+			expect = pb.Stage_Stopped
+		}
+	case *pb.StartTaskRequest, *pb.UpdateTaskRequest:
+		expect = pb.Stage_Running
+	case *pb.OperateTaskRequest:
+		req := masterReq.(*pb.OperateTaskRequest)
+		switch req.Op {
+		case pb.TaskOp_Resume:
+			expect = pb.Stage_Running
+		case pb.TaskOp_Pause:
+			expect = pb.Stage_Paused
+		case pb.TaskOp_Stop:
+			expect = pb.Stage_Stopped
+		}
+	case *pb.OperateWorkerRelayRequest:
+		req := masterReq.(*pb.OperateWorkerRelayRequest)
+		switch req.Op {
+		case pb.RelayOp_ResumeRelay:
+			expect = pb.Stage_Running
+		case pb.RelayOp_PauseRelay:
+			expect = pb.Stage_Paused
+		case pb.RelayOp_StopRelay:
+			expect = pb.Stage_Stopped
+		}
+	default:
+		return nil, terror.ErrMasterIsNotAsyncRequest.Generate(masterReq)
+	}
+	req := &workerrpc.Request{
+		Type: workerrpc.CmdQueryStatus,
+		QueryStatus: &pb.QueryStatusRequest{
+			Name: taskName,
+		},
 	}
 
-	worker := s.scheduler.GetWorkerBySource(cfg.SourceID)
-	if worker == nil {
-		return nil, "", "", handleErr(terror.ErrMasterTaskConfigExtractor.Generatef("%s relevant worker-client not found", cfg.SourceID))
+	for num := 0; num < maxRetryNum; num++ {
+		// check whether source relative worker has been removed by scheduler
+		switch masterReq.(type) {
+		case *pb.OperateSourceRequest:
```

Comment:
seems that no need to use `switch`

Reply:
revised in https://github.com/pingcap/dm/pull/490/commits/6d43a1ddce1f583c1a734b38e787d65cc9ad6261